datafusion/physical-expr/src/datetime_expressions.rs

```diff
@@ -497,6 +502,99 @@ pub fn make_current_time(
     move |_arg| Ok(ColumnarValue::Scalar(ScalarValue::Time64Nanosecond(nano)))
 }
 
+/// make_date(year, month, date) SQL function implementation
+pub fn make_date(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+    if args.len() != 3 {
+        return exec_err!(
+            "make_date function requires 3 arguments, got {}",
+            args.len()
+        );
+    }
+
+    // first, identify if any of the arguments is an Array. If yes, store its `len`,
+    // as any scalar will need to be converted to an array of len `len`.
+    let len = args
+        .iter()
+        .fold(Option::<usize>::None, |acc, arg| match arg {
+            ColumnarValue::Scalar(_) => acc,
+            ColumnarValue::Array(a) => Some(a.len()),
+        });
+
+    let is_scalar = len.is_none();
+    let array_size = if is_scalar { 1 } else { len.unwrap() };
+
+    let years = cast_column(&args[0], &DataType::Int32, None)?;
+    let months = cast_column(&args[1], &DataType::Int32, None)?;
+    let days = cast_column(&args[2], &DataType::Int32, None)?;
+
+    let value_fn = |col: &ColumnarValue, pos: usize| -> Result<i32> {
+        match col {
+            ColumnarValue::Array(a) => Ok(a.as_primitive::<Int32Type>().value(pos)),
+            ColumnarValue::Scalar(s) => match s {
+                ScalarValue::Int32(i) => match i {
+                    Some(i) => Ok(*i),
+                    None => {
+                        exec_err!("Unable to parse date from null/empty value")
+                    }
+                },
+                _ => unreachable!(),
+            },
+        }
+    };
+
+    // since the epoch for the date32 datatype is the unix epoch
+    // we need to subtract the unix epoch from the current date
+    // note this can result in a negative value
+    let unix_days_from_ce = NaiveDate::from_ymd_opt(1970, 1, 1)
+        .unwrap()
+        .num_days_from_ce();
+
+    let mut builder: PrimitiveBuilder<Date32Type> = PrimitiveArray::builder(array_size);
+
+    for i in 0..array_size {
+        let y = value_fn(&years, i)?;
+        let m = u32::try_from(value_fn(&months, i)?);
+        let d = u32::try_from(value_fn(&days, i)?);
+
+        if m.is_err() {
+            return exec_err!(
+                "Month value '{:?}' is out of range",
+                value_fn(&months, i).unwrap()
+            );
+        }
```

Comment:
I think you can write this logic like so to avoid the unwraps below:

```suggestion
        let Ok(m) = m else {
            return exec_err!(
                "Month value '{:?}' is out of range",
                value_fn(&months, i).unwrap()
            );
        };
```

Reply:
Proposed PR to do this in https://github.com/apache/arrow-datafusion/pull/9072